lib/index.ts

```diff
@@ -263,9 +263,13 @@ export async function performBuilds(
 		secrets: secretMap,
 		regSecrets: registrySecrets,
 		architecture,
+		experimental,
 	} = await initializeBuildMetadata(tasks, docker, tmpDir);
 
 	const images = await Bluebird.map(tasks, (task: BuildTask) => {
+		if (experimental) {
```

Comment:
But what if an upcoming (or recently released?!) Docker version "promotes" the `platform` build option from experimental to standard? In fact, in Docker engine API v1.40, the `platform` option is not labeled experimental:
https://docs.docker.com/engine/api/v1.40/#operation/ImageBuild
(Unlike other options like "squash".)

If this happened, balena users' builds could suddenly start failing with "exec format error" simply because the updated Docker on their laptops to a newer version that "fully supports" the `platform` option in a non-experimental way.

What is the harm of adding the `platform` option to `task.dockerOpts`, if the engine is older and does not understand it? I would hope that the engine would simply ignore the option, but if it actually fails the build, then that's an argument for having the experimental test.

Reply:
good point. we should be able to make that dependent on the api version as well I guess

Reply:
@robertgzr, this `if (experimental)` test was removed as discussed, so that the image build API request always includes the `platform` option, despite the API documentation saying that it requires API version >= 1.34. We discussed that we should then test with Docker v17, v18, v19 on Windows / macOS / Linux, to see if the `balena build` command would result in error (from the Docker Engine). I have tested only on macOS, and got a somewhat unexpected result.

Up to Docker Desktop / Engine version 18.03.1 **(API version 1.37, Experimental: true),** the build failed with the following error:

```
$ docker version
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:13:02 2018
 OS/Arch:      darwin/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:22:38 2018
  OS/Arch:      linux/amd64
  Experimental: true

$ ./bin/balena-dev build ~/balena/other/simple_container/ -A armv7hf -d raspberrypi3 --pull --nocache
...
[Error]   Build failed.
(HTTP code 400) unexpected - invalid platform: invalid platform architecture "arm"
```

The next version of Docker Desktop for macOS available for download is 18.06.0 **(API version 1.38, Experimental: true),** which succeeds:

```
$ docker version
Client:
 Version:           18.06.0-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        0ffa825
 Built:             Wed Jul 18 19:05:26 2018
 OS/Arch:           darwin/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:13:46 2018
  OS/Arch:          linux/amd64
  Experimental:     true

$ ./bin/balena-dev build ~/balena/other/simple_container/ -A armv7hf -d raspberrypi3 --pull --nocache
...
[Build]   Built 1 service in 0:27
[Success] Build succeeded!
```

So my tentative conclusion is that we need to add a test, but not a test for Experimental nor for the API version (given that `Experimental: true` and API v1.37 fail to build), but rather a test for the Docker Engine version, which should be `>= 18.06.0`. Perhaps we should try testing on Linux as well, to further clarify this conclusion.

Reply:
I have reproduced the results on Linux as well -- sort of, using docker-in-docker images. The build succeeds with Docker Engine v18.06.0 and later, and fails (`(HTTP code 400) unexpected - invalid platform: invalid platform architecture "arm"`) with v18.05.0 and earlier, even when `Experimental: false`.

**`balena build` fails with Docker v18.05.0 or earlier** (when the `platform` option is provided to the Docker Engine)
```
$ docker run --privileged --rm --name dind-18.05.0-ce -p 127.0.0.1:58992:2375 docker:18.05.0-ce-dind
...
time="2020-10-05T23:31:36.285042723Z" level=info msg="API listen on [::]:2375"

$ DOCKER_HOST=tcp://127.0.0.1:58992 docker version
...
Server:
 Engine:
  Version:          18.05.0-ce
  API version:      1.37 (minimum version 1.12)
  Go version:       go1.10.1
  Git commit:       f150324
  Built:            Wed May  9 22:20:42 2018
  OS/Arch:          linux/amd64
  Experimental:     false

$ DOCKER_HOST=tcp://127.0.0.1:58992 ./bin/balena-dev build ~/balena/other/simple_container/ -A armv7hf -d raspberrypi3 --pull --nocache
...
[Error]   Build failed.
(HTTP code 400) unexpected - invalid platform: invalid platform architecture "arm"
```

**`balena build` succeeds with Docker v18.06.0 or later** (when the `platform` option is provided to the Docker Engine)
```
$ docker run --privileged --rm --name dind-18.06.0-ce -p 127.0.0.1:58992:2375 docker:18.06.0-ce-dind
...
time="2020-10-05T23:32:21.139320655Z" level=info msg="API listen on [::]:2375"

$ DOCKER_HOST=tcp://127.0.0.1:58992 docker version
...
Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:13:39 2018
  OS/Arch:          linux/amd64
  Experimental:     false

 $ DOCKER_HOST=tcp://127.0.0.1:58992 ./bin/balena-dev build ~/balena/other/simple_container/ -A armv7hf -d raspberrypi3 --pull --nocache
...
[Build]   Built 1 service in 0:23
[Success] Build succeeded!
```

@robertgzr, @tmigone, it sounds to me like this PR should test for Docker Engine version 18.06.0 or later to decide whether or not to include the `platform` option in the docker options provided to the Docker Engine. I have not tested with balenaEngine, I assume it would be the same...

Reply:
Note that we don't have any check on the current revision. It looks like you're testing on an amd64 host, for that case we will always need to use emulation, which is what we did before right?
If the user has `binfmt_misc` installed that won't be strictly necessary but I guess we would want to use qemu anyway....

Reply:
> It looks like you're testing on an amd64 host, for that case we will always need to use emulation, which is what we did before right?

Yes. Adding the `--emulated` option to `balena build` makes no difference here though (I tested), because the error message:
```
(HTTP code 400) unexpected - invalid platform: invalid platform architecture "arm"
```
... appears to happen _before_ the Docker Engine starts building the image. The Docker Engine rejects the image build API request, as if during request validation: https://github.com/balena-os/balena-engine/blob/v18.04.0-ce-rc1/pkg/system/lcow.go#L15-L21

Indeed, HTTP code 400 means "Bad Request".

Reply:
the thing is on balena engine 18.09.x and 19.03.x we should have the check for windows:
https://github.com/balena-os/balena-engine/blob/18.09-balena/pkg/system/lcow.go#L25
https://github.com/balena-os/balena-engine/blob/19.03-balena/pkg/system/lcow.go#L25

it does look like we need to check for version `18.06.0`, specifically because of this commit:
https://github.com/moby/moby/commit/7f334d3acfd7bfde900e16e393662587b9ff74a1

Reply:
In that case let's restrict this to api version `>=1.38` which should get us out of trouble

Reply:
I think with the latest commit we should be good